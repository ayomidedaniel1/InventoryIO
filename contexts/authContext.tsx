import React, {
  useState,
  useEffect,
  useContext,
  createContext,
  FC,
  ReactNode,
} from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

export type User = {
  email: string;
};

type AuthContextType = {
  user: User | null;
  isLoggedIn: boolean;
  loading: boolean;
  login: (user: User) => void;
  logout: () => void;
};

type AuthProviderProps = {
  children: ReactNode;
};

export const AuthContext = createContext<AuthContextType>(
  undefined as never
);

export const AuthProvider: FC<AuthProviderProps> = ({ children }) => {
  const [user, setUser] = useState<User | null>(null);
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    checkUser();
  }, []);

  const checkUser = async () => {
    setLoading(true);
    try {
      const isUserLoggedIn = await AsyncStorage.getItem('isLoggedIn');
      const storedUser = await AsyncStorage.getItem('user');

      if (isUserLoggedIn !== null && storedUser) {
        setIsLoggedIn(true);
        setUser(JSON.parse(storedUser));
      }
    } catch (error) {
      console.error("Error retrieving user ", error);
    } finally {
      setLoading(false);
    }
  };

  const login = (newUser: User) => {
    AsyncStorage.setItem("isLoggedIn", 'true');
    AsyncStorage.setItem("user", JSON.stringify(newUser));
    setIsLoggedIn(true);
    setUser(newUser);
  };

  const logout = () => {
    setUser(null);
    AsyncStorage.removeItem("isLoggedIn");
    AsyncStorage.removeItem("user");
    setIsLoggedIn(false);
  };

  const contextValue: AuthContextType = {
    user,
    loading,
    isLoggedIn,
    login,
    logout,
  };

  return (
    <AuthContext.Provider value={contextValue}>
      {children}
    </AuthContext.Provider>
  );
};

export function useAuthContext(): AuthContextType {
  return useContext(AuthContext);
}