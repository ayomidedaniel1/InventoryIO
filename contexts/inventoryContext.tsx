import React, {
  useState,
  useEffect,
  useContext,
  createContext,
  FC,
  ReactNode,
} from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { useAuthContext } from "./authContext";

type InventoryItem = {
  name: string;
  totalStock: number;
  price: number;
  description: string;
};

type InventoryContextType = {
  inventory: InventoryItem[];
  addInventory: (item: InventoryItem) => void;
  editInventory: (name: string, newItem: InventoryItem) => void;
  deleteInventory: (name: string) => void;
};

type InventoryProviderProps = {
  children: ReactNode;
};

export const InventoryContext = createContext<InventoryContextType | null>(null);

export const InventoryProvider: FC<InventoryProviderProps> = ({ children }) => {
  const authContext = useAuthContext();
  const [inventory, setInventory] = useState<InventoryItem[]>([]);

  useEffect(() => {
    const loadInventory = async () => {
      try {
        const storedInventory = await AsyncStorage.getItem('inventory');

        if (storedInventory) {
          setInventory(JSON.parse(storedInventory));
        }
      } catch (error) {
        console.error("Error loading inventory ", error);
      }
    };

    loadInventory();
  }, []);

  const addInventory = async (item: InventoryItem) => {
    const newInventory = [...inventory, item];
    setInventory(newInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(newInventory));
  };

  const editInventory = async (name: string, newItem: InventoryItem) => {
    const tempInventory = [...inventory];

    const updatedInventory = tempInventory.map(item => {
      return item.name === name ? newItem : item;
    });
    setInventory(updatedInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(updatedInventory));
  };

  const deleteInventory = async (name: string) => {
    const updatedInventory = inventory.filter(item => item.name !== name);
    setInventory(updatedInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(updatedInventory));
  };

  if (!authContext.isLoggedIn) {
    return null;
  }

  const contextValue: InventoryContextType = {
    inventory,
    addInventory,
    editInventory,
    deleteInventory,
  };

  return (
    <InventoryContext.Provider value={contextValue}>
      {children}
    </InventoryContext.Provider>
  );
};

export function useInventoryContext(): InventoryContextType {
  const context = useContext(InventoryContext);
  if (!context) {
    throw new Error("Some error occurred");
  }
  return context;
}