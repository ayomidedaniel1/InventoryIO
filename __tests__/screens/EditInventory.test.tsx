import React from "react";
import renderer from 'react-test-renderer';
import { InventoryProvider } from "../../contexts/inventoryContext";
import { AuthProvider } from "../../contexts/authContext";

jest.mock('react-native-simple-toast', () => ({
  show: jest.fn(),
}));

import EditInventoryScreen from "../../app/(public)/editInventory";

test('Edit screen renders correctly', () => {
  const tree = renderer.create(
    <AuthProvider>
      <InventoryProvider>
        <EditInventoryScreen />
      </InventoryProvider>
    </AuthProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
