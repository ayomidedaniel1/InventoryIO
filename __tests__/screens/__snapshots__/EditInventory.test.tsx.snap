// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit screen renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFF",
      "flex": 1,
      "paddingTop": 60,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#6c47ff",
        "height": 80,
        "marginBottom": 50,
        "marginTop": -50,
        "width": 750,
      }
    }
  />
  <View
    style={
      {
        "position": "absolute",
        "top": 50,
        "width": 750,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "position": "relative",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 100,
          }
        }
      >
        <Text />
      </View>
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 24,
            "fontStyle": "italic",
            "fontWeight": "800",
            "marginRight": 20,
            "textAlign": "center",
          }
        }
      >
        Edit Inventory
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#606060",
        "fontSize": 16,
        "fontWeight": "600",
        "marginBottom": 10,
        "marginTop": -20,
        "textAlign": "center",
      }
    }
  >
    Edit this item in your inventory
  </Text>
  <View
    style={
      {
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 8,
          "marginHorizontal": 20,
          "marginTop": 14,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000",
            "fontWeight": "600",
            "paddingBottom": 8,
          }
        }
      >
        Item Name
      </Text>
      <TextInput
        defaultValue=""
        keyboardType="default"
        onChangeText={[Function]}
        placeholder="Your item name"
        placeholderTextColor="#A6A6A6"
        selectionColor="#A6A6A6"
        style={
          {
            "backgroundColor": "#F9F9F9",
            "borderColor": "#CCC9C9",
            "borderRadius": 6,
            "borderWidth": 1,
            "color": "#A6A6A6",
            "fontSize": 15,
            "fontWeight": "500",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginBottom": 8,
          "marginHorizontal": 20,
          "marginTop": 14,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000",
            "fontWeight": "600",
            "paddingBottom": 8,
          }
        }
      >
        Total Stock
      </Text>
      <TextInput
        defaultValue=""
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Your item's total stock"
        placeholderTextColor="#A6A6A6"
        selectionColor="#A6A6A6"
        style={
          {
            "backgroundColor": "#F9F9F9",
            "borderColor": "#CCC9C9",
            "borderRadius": 6,
            "borderWidth": 1,
            "color": "#A6A6A6",
            "fontSize": 15,
            "fontWeight": "500",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginBottom": 8,
          "marginHorizontal": 20,
          "marginTop": 14,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000",
            "fontWeight": "600",
            "paddingBottom": 8,
          }
        }
      >
        Item Price
      </Text>
      <TextInput
        defaultValue=""
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Your item's price"
        placeholderTextColor="#A6A6A6"
        selectionColor="#A6A6A6"
        style={
          {
            "backgroundColor": "#F9F9F9",
            "borderColor": "#CCC9C9",
            "borderRadius": 6,
            "borderWidth": 1,
            "color": "#A6A6A6",
            "fontSize": 15,
            "fontWeight": "500",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginBottom": 8,
          "marginHorizontal": 20,
          "marginTop": 14,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000",
            "fontWeight": "600",
            "paddingBottom": 8,
          }
        }
      >
        Item Description
      </Text>
      <TextInput
        defaultValue=""
        keyboardType="default"
        onChangeText={[Function]}
        placeholder="Describe your item"
        placeholderTextColor="#A6A6A6"
        selectionColor="#A6A6A6"
        style={
          {
            "backgroundColor": "#F9F9F9",
            "borderColor": "#CCC9C9",
            "borderRadius": 6,
            "borderWidth": 1,
            "color": "#A6A6A6",
            "fontSize": 15,
            "fontWeight": "500",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#0386D0",
          "borderRadius": 27,
          "height": 45,
          "justifyContent": "center",
          "marginTop": 52,
          "opacity": 1,
          "width": 151,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 14,
          }
        }
      >
        Save
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "red",
          "borderRadius": 6,
          "height": 32,
          "justifyContent": "center",
          "marginTop": 32,
          "opacity": 1,
          "width": 112,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 14,
          }
        }
      >
        Delete Item
      </Text>
    </View>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
